### DISCUSSION TITLE
task里面无法用get_config获取配置

### DISCUSSION THREAD

**Discussion by fripSide** [xmake-io/xmake#2182] [Q&A]
### Xmake 版本

xmake v2.6.4+HEAD.9361ffcc1,

### 操作系统版本和架构

Windows 10

### 描述问题

我写了这样一个task，在编译阶段合并bin文件，要调用sdk
```
task("link-all")
        set_category("plugin")
	on_run(function ()
		print("run task......")
		print(get_config("sdk"))
		local binutil = get_config("sdk").."/bin/arm-none-eabi-"  -- <-----------------这里无法获取sdk
               -- ....
       end)
       set_menu -- ....
task_end()
```
如果直接执行这个task，或者在另一个任务调用这个task，就无法获取sdk。
error: attempt to concatenate a nil value

只能在target里面调用这个task。

```
after_build(function (target)
  -- import task module
  import("core.project.task")
  task.run("link-all")
end)
```

**另外：这个功能我本来是写在rule里面。不同的rule的after_build为什么不能保证执行顺序？link-all依赖arm-gcc的结果。并且task下面的after_build和add_rules里面的after_build也无法保证顺序。。。**
``` 
target(app_name)
	set_kind("binary")
	add_rules("arm-gcc")
        add_rules("link-all")
```

### 期待的结果

在任何地方都应该能get_config，不应该是nil。因为我在第二行就import了toolchian

```
set_toolchains("arm-gcc")
```

### 附加信息和错误日志

> xmake link-all
error: attempt to concatenate a nil value

**Comment by waruqi** [comment]
> 如果直接执行这个task，或者在另一个任务调用这个task，就无法获取sdk。

正常，单独定义执行 task，就是个完整的插件，独立于工程的存在，即使没有任何工程，也可以执行 xmake xxx 的 独立task

而 get_config 仅仅只是工程相关的本地配置读取。。

而你在 after_build 调用 task，因为已经加载了工程配置，当然可以读到了。。

单独执行，要么用 task 定义自己的 options, `--sdk`，然后单独传递 xmake xxx --sdk=xxxx，在 task 里面通过 import core.base.option 模块获取，option.get("sdk")

要么，import core.project.config 模块，然后先 config.load() 加载本地工程配置，这个时候 get_config 就能取到，但是这种有限制，首次执行 task，你肯定取不到，因为还没执行 `xmake f/config --sdk=`，本地配置还没缓存。

> 这个功能我本来是写在rule里面。不同的rule的after_build为什么不能保证执行顺序？link-all依赖arm-gcc的结果。并且task下面的after_build和add_rules里面的after_build也无法保证顺序。。。

rule 之间通过 rule 的 add_deps 保证执行顺序，否则就是同级的，并行执行，顺序不保证。

**Reply by fripSide** [reply]
为什么项目之间也不保证build顺序。我的项目结构如下：  

```
target1()
target2() 
        task2
```

我有个task2需要在target1,target2（都是binary）完成之后再执行。目前是放在target2的rule里面。怎么保证target1比target2先build。

**Reply by waruqi** [reply]
同样，走 add_deps 控制顺序，否则无依赖，可并行执行，不保证顺序

**Reply by fripSide** [reply]
能否定义一个所有target build都结束的命令，例如：project:after_build

**Reply by waruqi** [reply]
你定义个 phony target + build_after，通过 add_deps 依赖其他所有 target 不就好了

**Reply by fripSide** [reply]
我目前是添加依赖，这样能解决问题:
target2:
     add_deps("target1")

target1,2都是binary，很难猜出有这种用法。之前以为add_deps只能添加static库的依赖。文档上面的确说了决定build顺序。

**Reply by waruqi** [reply]
> 我目前是添加依赖，这样能解决问题: target2: add_deps("target1")
> 
> target1,2都是binary，很难猜出有这种用法。

这是最基础的用法，文档里面 都有说明，不是靠猜的，得看文档  https://xmake.io/#/zh-cn/manual/project_target?id=targetadd_deps
